src/main/scala/wop/game/ai/WopSolver.scala

```diff
@@ -96,43 +104,51 @@ object WopSolver {
       case Player.P1 => gt(score, beta)
       case Player.P2 => lt(score, alpha)
     }
+
+    def playerFactor(xo: WopState.XO): Int = xo match {
+      case WopState.Player.P1.xo => 1
+      case WopState.Player.P2.xo => -1
+      case WopState.XO.Empty => 0
+    }
   }
 
-  def alphaBeta(state: Either[Foul, WopState],
-                prevPoint: Option[Point],
-                depth: Int = 0,
-                ab: Option[(Int, Int)] = None): EvalPoint = state match {
-    case Right(s: WopState) =>
-      lazy val returnResult = (WopSolver.heuristicEvalState(s), prevPoint)
-      if (depth >= MAX_DEPTH) returnResult
-      else
-        s match {
-          case is: InProgress =>
-            val alphaBetaSolver = AlphaBetaSolver(is.player, ab)
-            @tailrec def processChild(points: List[Point], player: Player, bestEval: EvalPoint): EvalPoint =
-              points match {
-                case (point: Point) :: tl =>
-                  val nextPoint = prevPoint match {
-                    case None => Some(point)
-                    case _ => prevPoint
+  def minMax(state: WopState): Option[Point] = {
+    def loop(state: Either[Foul, WopState], currPoint: Option[Point], depth: Int, ab: Option[(Int, Int)]): EvalPoint =
```

Comment:
minMax is not recursive, it is only interface wrapper of loop
and loop can't tailrec optimize